File: server/services/celestialObjects.ts
import { CelestialObject, InsertCelestialObject, InsertApodCache, apodCache } from "@shared/schema";
import { storage } from "../storage";
import { db } from "../db";
import { get2MASSImageUrl, isMessierObject } from "../data/messier2mass";

// Seed data for celestial objects (static catalog - no time-specific info)
const seedCelestialObjects: InsertCelestialObject[] = [
  {
    name: "Whirlpool Galaxy (M51)",
    type: "galaxy",
    description: "One of the most popular targets for visual observers and astrophotographers. Through a small scope, the galaxy appears as a faint patch with its companion NGC 5195. Larger telescopes (250mm+) show its famous spiral arms.",
    imageUrl: "https://www.ipac.caltech.edu/2mass/gallery/m51atlas.jpg",
    constellation: "Canes Venatici",
    magnitude: "8.4",
  },
  {
    name: "Leo Triplet (M65, M66, NGC 3628)",
    type: "galaxy",
    description: "A small group of galaxies about 35 million light-years away. M65 and M66 are visible with binoculars, but you'll need a scope for NGC 3628. At 70x, all three appear as elongated patches.",
    imageUrl: "https://upload.wikimedia.org/wikipedia/commons/8/83/Leo-Triplet.png",
    constellation: "Leo",
    magnitude: "9.3",
  },
  {
    name: "Mizar & Alcor",
    type: "double_star",
    description: "An outstanding double star for beginners in the handle of the Big Dipper. Mizar itself is double, making this a great target for any telescope.",
    imageUrl: "https://upload.wikimedia.org/wikipedia/commons/e/e4/ALCOR_et_MIZAR_%287060991417%29.jpg",
    constellation: "Ursa Major",
    magnitude: "2.2",
  },
  {
    name: "Ring Nebula (M57)",
    type: "nebula",
    description: "A planetary nebula formed by an expanding shell of gas around an aging star. Approximately 2,000 light-years away, it's one of the most prominent examples of its type.",
    imageUrl: "https://www.ipac.caltech.edu/2mass/gallery/ringatlas.jpg",
    constellation: "Lyra",
    magnitude: "8.8",
  },
  {
    name: "Jupiter",
    type: "planet",
    description: "The largest planet in our solar system. An 8-inch Dobsonian can reveal cloud bands and the four Galilean moons: Io, Europa, Ganymede, and Callisto.",
    imageUrl: "https://images-assets.nasa.gov/image/PIA22946/PIA22946~large.jpg",
    constellation: null,
    magnitude: "-2.2",
  },
  {
    name: "Saturn",
    type: "planet",
    description: "The ringed planet is a stunning sight in any telescope. Look for the Cassini Division in the rings and its largest moon Titan.",
    imageUrl: "https://images-assets.nasa.gov/image/PIA21046/PIA21046~large.jpg",
    constellation: null,
    magnitude: "0.5",
  },
  {
    name: "Andromeda Galaxy (M31)",
    type: "galaxy",
    description: "The nearest major galaxy to the Milky Way at 2.5 million light-years. Visible to the naked eye on dark nights, it's the largest galaxy in our Local Group.",
    imageUrl: "https://www.ipac.caltech.edu/2mass/gallery/m31atlas.jpg",
    constellation: "Andromeda",
    magnitude: "3.4",
  },
  {
    name: "Orion Nebula (M42)",
    type: "nebula",
    description: "A diffuse nebula south of Orion's Belt. One of the brightest nebulae, visible to the naked eye, and the closest region of massive star formation to Earth at 1,344 light-years.",
    imageUrl: "https://www.ipac.caltech.edu/2mass/gallery/orionatlas.jpg",
    constellation: "Orion",
    magnitude: "4.0",
  },
  {
    name: "Pleiades (M45)",
    type: "star_cluster",
    description: "The Seven Sisters - an open star cluster containing hot B-type stars. Contains over 1,000 confirmed members, best viewed with low power wide-field eyepieces.",
    imageUrl: "https://www.ipac.caltech.edu/2mass/gallery/pleiadesatlas.jpg",
    constellation: "Taurus",
    magnitude: "1.6",
  },
  {
    name: "Great Globular Cluster (M13)",
    type: "star_cluster",
    description: "One of the brightest globular clusters in the northern sky. Contains several hundred thousand stars packed into a sphere about 145 light-years across.",
    imageUrl: "https://www.ipac.caltech.edu/2mass/gallery/m13atlas.jpg",
    constellation: "Hercules",
    magnitude: "5.8",
  },
  {
    name: "Dumbbell Nebula (M27)",
    type: "nebula",
    description: "A large, bright planetary nebula. One of the easiest planetaries to observe, showing an hourglass or apple-core shape in modest telescopes.",
    imageUrl: "https://www.ipac.caltech.edu/2mass/gallery/dumbbellatlas.jpg",
    constellation: "Vulpecula",
    magnitude: "7.5",
  },
];

// Telescope tips seed data
const seedTelescopeTips = [
  {
    title: "Collimating Your Apertura AD8 Dobsonian",
    content: "Good collimation is crucial for sharp views with your Apertura AD8. The included laser collimator makes this easy with just two main steps: 1) First, align the secondary mirror by adjusting the secondary mirror housing hex screws until the laser hits the center spot on the primary mirror. 2) Then, align the primary mirror by loosening the white thumb screws and adjusting the black knobs until the laser returns back to the laser collimator's 45-degree reflective surface. This process takes under 2 minutes once you've practiced a few times and will significantly improve your viewing experience.",
    category: "Maintenance",
    imageUrl: "/collimate_AD8.jpg"
  },
  {
    title: "Best Eyepieces for Your Dob",
    content: "Discover which eyepieces work best with your 8-inch Dobsonian for different celestial objects, from planets to deep sky targets.",
    category: "Equipment",
    imageUrl: "https://images.unsplash.com/photo-1536697246787-1f7ae568d89a?auto=format&fit=crop&w=600&h=300"
  },
  {
    title: "Understanding Aperture",
    content: "When it comes to telescopes, there's one key feature that stands out from everything else: aperture. The aperture of a telescope is the diameter of the lens or mirror, and the bigger the aperture, the more light the telescope can gather. As a result, observers are able to identify fainter objects and see more detail than would be possible with a smaller aperture scope. The downside? Larger apertures can lack portability, and of course, they cost more!",
    category: "Astronomy Basics",
    imageUrl: "https://images.unsplash.com/photo-1522124624696-7ea32eb9592c?auto=format&fit=crop&w=600&h=300"
  }
];

/**
 * Seeds the database with initial celestial objects and telescope tips
 */
export async function seedDatabase(): Promise<void> {
  // Seed celestial objects
  const existingObjects = await storage.getAllCelestialObjects();
  if (existingObjects.length === 0) {
    for (const object of seedCelestialObjects) {
      await storage.createCelestialObject(object);
    }
    console.log('Seeded celestial objects');
  }

  // Seed telescope tips
  const existingTips = await storage.getAllTelescopeTips();
  if (existingTips.length === 0) {
    for (const tip of seedTelescopeTips) {
      await storage.createTelescopeTip(tip);
    }
    console.log('Seeded telescope tips');
  }

  // Check if we have an APOD cache entry
  try {
    const existingApodEntries = await db.select().from(apodCache);

    // If no APOD entries exist, seed with a default one
    if (existingApodEntries.length === 0) {
      const today = new Date();
      const year = today.getFullYear();
      const month = today.getMonth() + 1;
      const day = today.getDate();
      const formattedDate = `${year}-${month < 10 ? '0' + month : month}-${day < 10 ? '0' + day : day}`;

      const defaultApod: InsertApodCache = {
        date: formattedDate,
        title: "A Happy Sky over Bufa Hill in Mexico",
        explanation: "Sometimes, the sky itself seems to smile. A few days ago, visible over much of the world, an unusual superposition of our Moon with the planets Venus and Saturn created just such an iconic facial expression.",
        media_type: "image",
        service_version: "v1",
        url: "https://apod.nasa.gov/apod/image/2504/HappySkyMexico_Korona_960.jpg",
        hdurl: "https://apod.nasa.gov/apod/image/2504/HappySkyMexico_Korona_1358.jpg",
        copyright: "Daniel Korona"
      };

      await db.insert(apodCache).values(defaultApod);
      console.log("Seeded APOD cache with default entry");
    }
  } catch (error) {
    console.error("Error checking APOD cache:", error);
  }
}

/**
 * Gets the current month's name
 */
export function getCurrentMonth(): string {
  return new Date().toLocaleString('default', { month: 'long' });
}

/**
 * Gets the current year
 */
export function getCurrentYear(): number {
  return new Date().getFullYear();
}

/**
 * Get an appropriate image URL for a celestial object
 * Uses 2MASS for Messier objects, falls back to NASA search or type-specific defaults
 */
export function getImageUrlForObject(name: string, type: string): string | null {
  // Check if it's a Messier object first
  const messierUrl = get2MASSImageUrl(name);
  if (messierUrl) {
    return messierUrl;
  }

  // Return null - caller should use NASA image search or fallback
  return null;
}

/**
 * Get type-specific fallback image URLs
 */
export function getTypeSpecificFallbackImage(type: string): string {
  switch (type) {
    case 'galaxy':
      return 'https://images.unsplash.com/photo-1502134249126-9f3755a50d78?w=800&h=500&fit=crop';
    case 'nebula':
      return 'https://images.unsplash.com/photo-1462331940025-496dfbfc7564?w=800&h=500&fit=crop';
    case 'planet':
      return 'https://images.unsplash.com/photo-1614728263952-84ea256f9679?w=800&h=500&fit=crop';
    case 'star_cluster':
      return 'https://images.unsplash.com/photo-1419242902214-272b3f66ee7a?w=800&h=500&fit=crop';
    case 'double_star':
      return 'https://images.unsplash.com/photo-1502134249126-9f3755a50d78?w=800&h=500&fit=crop';
    case 'moon':
      return 'https://images.unsplash.com/photo-1446776877081-d282a0f896e2?w=800&h=500&fit=crop';
    default:
      return 'https://images.unsplash.com/photo-1462331940025-496dfbfc7564?w=800&h=500&fit=crop';
  }
}

File: python_server/services/celestial_objects.py
from datetime import datetime
from typing import List, Dict, Optional, Any
from data.storage import storage
from data.models import CelestialObject, MonthlyGuide, TelescopeTip

def seed_database() -> None:
    """
    Seeds the database with initial celestial objects, monthly guides, and telescope tips
    """
    # Check if we've already seeded the database
    if storage.celestial_objects:
        return
    
    # Sample celestial objects
    celestial_objects = [
        {
            "name": "Leo Triplet",
            "type": "galaxy",
            "description": "A small group of galaxies about 35 million light-years away in the constellation Leo.",
            "coordinates": "RA: 11h 18m | Dec: +13° 03′",
            "month": "April",
            "bestViewingTime": "9 PM - 3 AM",
            "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/thumb/a/ae/Leo_Triplet.jpg/600px-Leo_Triplet.jpg",
            "visibilityRating": "Good",
            "information": "Also known as the M66 Group, this trio includes the spiral galaxies M65, M66, and NGC 3628.",
            "constellation": "Leo",
            "magnitude": "9-10",
            "hemisphere": "Both",
            "recommendedEyepiece": "Low power"
        },
        {
            "name": "Jupiter",
            "type": "planet",
            "description": "The largest planet in our solar system, known for its Great Red Spot and numerous moons.",
            "coordinates": "Varies by date",
            "month": "August",
            "bestViewingTime": "10 PM - 4 AM",
            "imageUrl": "https://solarsystem.nasa.gov/system/stellar_items/image_files/6_jupiter.jpg",
            "visibilityRating": "Excellent",
            "information": "With a good telescope, you can observe Jupiter's cloud bands and its four largest moons: Io, Europa, Ganymede, and Callisto.",
            "constellation": "Varies by season",
            "magnitude": "-2.7",
            "hemisphere": "Both",
            "recommendedEyepiece": "Medium power (100-150x)"
        },
        {
            "name": "M13 (Great Globular Cluster)",
            "type": "star_cluster",
            "description": "One of the most prominent and best-known globular clusters in the northern sky.",
            "coordinates": "RA: 16h 41m | Dec: +36° 28′",
            "month": "June",
            "bestViewingTime": "11 PM - 2 AM",
            "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/thumb/6/6a/Messier_13.jpg/600px-Messier_13.jpg",
            "visibilityRating": "Good",
            "information": "Contains about 300,000 stars and is approximately 22,000 light-years away from Earth.",
            "constellation": "Hercules",
            "magnitude": "5.8",
            "hemisphere": "Northern",
            "recommendedEyepiece": "Medium power"
        },
        {
            "name": "Orion Nebula (M42)",
            "type": "nebula",
            "description": "A diffuse nebula situated in the Milky Way, south of Orion's Belt.",
            "coordinates": "RA: 05h 35m | Dec: -05° 23′",
            "month": "January",
            "bestViewingTime": "8 PM - 12 AM",
            "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/thumb/f/f3/Orion_Nebula_-_Hubble_2006_mosaic_18000.jpg/600px-Orion_Nebula_-_Hubble_2006_mosaic_18000.jpg",
            "visibilityRating": "Excellent",
            "information": "One of the brightest nebulae and visible to the naked eye under dark skies.",
            "constellation": "Orion",
            "magnitude": "4.0",
            "hemisphere": "Both",
            "recommendedEyepiece": "Low power"
        },
        {
            "name": "Saturn",
            "type": "planet",
            "description": "The sixth planet from the Sun, known for its prominent ring system.",
            "coordinates": "Varies by date",
            "month": "July",
            "bestViewingTime": "10 PM - 3 AM",
            "imageUrl": "https://solarsystem.nasa.gov/system/stellar_items/image_files/38_saturn_1600x900.jpg",
            "visibilityRating": "Excellent",
            "information": "Saturn's rings are visible with even a small telescope, making it one of the most rewarding objects for amateur astronomers.",
            "constellation": "Varies by season",
            "magnitude": "0.1",
            "hemisphere": "Both",
            "recommendedEyepiece": "Medium to high power (150-200x)"
        },
        {
            "name": "Andromeda Galaxy (M31)",
            "type": "galaxy",
            "description": "The nearest major galaxy to the Milky Way, visible to the naked eye on moonless nights.",
            "coordinates": "RA: 00h 42m | Dec: +41° 16′",
            "month": "October",
            "bestViewingTime": "9 PM - 2 AM",
            "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/thumb/9/98/Andromeda_Galaxy_%28with_h-alpha%29.jpg/600px-Andromeda_Galaxy_%28with_h-alpha%29.jpg",
            "visibilityRating": "Good",
            "information": "Located approximately 2.5 million light-years from Earth, it is the most distant object visible to the naked eye.",
            "constellation": "Andromeda",
            "magnitude": "3.4",
            "hemisphere": "Northern",
            "recommendedEyepiece": "Low power"
        },
        {
            "name": "Pleiades (M45)",
            "type": "star_cluster",
            "description": "An open star cluster containing middle-aged, hot B-type stars located in the constellation of Taurus.",
            "coordinates": "RA: 03h 47m | Dec: +24° 07′",
            "month": "November",
            "bestViewingTime": "8 PM - 1 AM",
            "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/thumb/4/4e/Pleiades_large.jpg/600px-Pleiades_large.jpg",
            "visibilityRating": "Excellent",
            "information": "Also known as the Seven Sisters, this cluster is among the nearest star clusters to Earth at just 444 light-years away.",
            "constellation": "Taurus",
            "magnitude": "1.6",
            "hemisphere": "Both",
            "recommendedEyepiece": "Low power or binoculars"
        },
        {
            "name": "Whirlpool Galaxy (M51)",
            "type": "galaxy",
            "description": "A grand-design spiral galaxy located in the constellation Canes Venatici.",
            "coordinates": "RA: 13h 29m | Dec: +47° 11′",
            "month": "April",
            "bestViewingTime": "10 PM - 2 AM",
            "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/thumb/d/db/Messier51_sRGB.jpg/600px-Messier51_sRGB.jpg",
            "visibilityRating": "Moderate",
            "information": "The galaxy and its companion, NGC 5195, are easily observed by amateur astronomers, and the two galaxies may be seen with binoculars.",
            "constellation": "Canes Venatici",
            "magnitude": "8.4",
            "hemisphere": "Northern",
            "recommendedEyepiece": "Medium power"
        },
        {
            "name": "Crab Nebula (M1)",
            "type": "nebula",
            "description": "A supernova remnant and pulsar wind nebula in the constellation of Taurus.",
            "coordinates": "RA: 05h 34m | Dec: +22° 01′",
            "month": "December",
            "bestViewingTime": "8 PM - 11 PM",
            "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/thumb/0/00/Crab_Nebula.jpg/600px-Crab_Nebula.jpg",
            "visibilityRating": "Moderate",
            "information": "Corresponds to a bright supernova recorded by Chinese astronomers in 1054 AD.",
            "constellation": "Taurus",
            "magnitude": "8.4",
            "hemisphere": "Both",
            "recommendedEyepiece": "Medium power"
        },
        {
            "name": "Albireo",
            "type": "double_star",
            "description": "A beautiful double star and the fifth-brightest in the constellation Cygnus.",
            "coordinates": "RA: 19h 30m | Dec: +27° 57′",
            "month": "July",
            "bestViewingTime": "10 PM - 1 AM",
            "imageUrl": "https://upload.wikimedia.org/wikipedia/commons/c/c2/Albireo_Lund_2020.jpg",
            "visibilityRating": "Good",
            "information": "Considered one of the most beautiful double stars, with a gold primary and blue-green secondary component.",
            "constellation": "Cygnus",
            "magnitude": "3.1",
            "hemisphere": "Northern",
            "recommendedEyepiece": "Medium to high power"
        }
    ]
    
    # Add the celestial objects to storage
    for obj_data in celestial_objects:
        storage.create_celestial_object(obj_data)
    
    print("Seeded celestial objects")
    
    # Monthly guides
    monthly_guides = [
        {
            "month": "April",
            "year": 2025,
            "headline": "Galaxy Season Peaks",
            "content": "April is an excellent time for observing galaxies as the night sky is focused away from the dense star clouds of the Milky Way, allowing us to peer into the deeper universe. The Leo Triplet and Whirlpool Galaxy are at their best this month.",
            "hemisphere": "Northern",
            "featuredObjects": [1, 8]  # IDs of Leo Triplet and Whirlpool Galaxy
        },
        {
            "month": "June",
            "year": 2025,
            "headline": "The Summer Sky Emerges",
            "content": "As summer begins, Hercules rises high in the night sky, bringing with it M13, one of the finest globular clusters visible from the northern hemisphere. Look for it in the 'keystone' of Hercules.",
            "hemisphere": "Northern",
            "featuredObjects": [3]  # ID of M13
        },
        {
            "month": "July",
            "year": 2025,
            "headline": "Planets and Double Stars",
            "content": "July brings excellent views of Saturn, which is approaching opposition, providing a perfect chance to observe its magnificent rings. Meanwhile, Albireo, a beautiful double star with contrasting colors, is high in the summer sky.",
            "hemisphere": "Northern",
            "featuredObjects": [5, 10]  # IDs of Saturn and Albireo
        },
        {
            "month": "October",
            "year": 2025,
            "headline": "Andromeda Nights",
            "content": "October offers dark autumn skies perfect for viewing the Andromeda Galaxy (M31), our closest major galactic neighbor. Under dark skies, it's visible to the naked eye as a misty patch in the constellation Andromeda.",
            "hemisphere": "Northern",
            "featuredObjects": [6]  # ID of Andromeda Galaxy
        },
        {
            "month": "December",
            "year": 2025,
            "headline": "Winter's Deep Sky Gems",
            "content": "The winter sky offers some of the most brilliant deep sky objects. The Crab Nebula (M1) is well-placed for evening observing, and the Pleiades (M45) is a stunning sight in binoculars or a small telescope.",
            "hemisphere": "Northern",
            "featuredObjects": [7, 9]  # IDs of Pleiades and Crab Nebula
        }
    ]
    
    # Add the monthly guides to storage
    for guide_data in monthly_guides:
        storage.create_monthly_guide(guide_data)
    
    print("Seeded monthly guides")
    
    # Telescope tips
    telescope_tips = [
        {
            "title": "Collimation Guide",
            "content": "Proper alignment of your telescope's optical elements (collimation) is essential for achieving sharp, high-contrast views. For Dobsonian owners, check your collimation before each observing session using a collimation cap or Cheshire eyepiece.",
            "category": "maintenance",
            "imageUrl": "https://images.unsplash.com/photo-1548639136-5ebdb06a6614?auto=format&fit=crop&w=300&h=200"
        },
        {
            "title": "Light Pollution Filters",
            "content": "Consider investing in a narrowband light pollution filter, which can significantly improve contrast on nebulae by blocking light pollution while allowing the light from emission nebulae to pass through.",
            "category": "equipment",
            "imageUrl": "https://images.unsplash.com/photo-1519638831568-d9897f54ed69?auto=format&fit=crop&w=300&h=200"
        },
        {
            "title": "Dark Adaptation",
            "content": "Allow your eyes at least 20-30 minutes to fully adapt to darkness. Use a red flashlight for reading star charts or adjusting equipment, as red light preserves your night vision.",
            "category": "observing",
            "imageUrl": "https://images.unsplash.com/photo-1533162507191-d90c625b2640?auto=format&fit=crop&w=300&h=200"
        },
        {
            "title": "Magnification Tips",
            "content": "Remember that highest power isn't always best. For most objects, moderate magnification with good contrast will provide better views than pushing your telescope to its magnification limits.",
            "category": "observing",
            "imageUrl": "https://images.unsplash.com/photo-1566904312366-0b95e8228de1?auto=format&fit=crop&w=300&h=200"
        },
        {
            "title": "Dew Prevention",
            "content": "In humid conditions, dew can form quickly on your telescope's optical surfaces. A dew shield or battery-powered dew heater can prevent this issue and extend your observing time.",
            "category": "equipment",
            "imageUrl": "https://images.unsplash.com/photo-1543722530-d2c3201371e7?auto=format&fit=crop&w=300&h=200"
        }
    ]
    
    # Add the telescope tips to storage
    for tip_data in telescope_tips:
        storage.create_telescope_tip(tip_data)
    
    print("Seeded telescope tips")

def get_current_month() -> str:
    """
    Gets the current month's name
    """
    return datetime.now().strftime('%B')

def get_current_year() -> int:
    """
    Gets the current year
    """
    return datetime.now().year

def filter_celestial_objects(object_type: Optional[str] = None, month: Optional[str] = None, hemisphere: Optional[str] = None) -> List[Dict[str, Any]]:
    """
    Filters celestial objects by multiple criteria
    """
    objects = storage.get_all_celestial_objects()
    filtered_objects = objects
    
    if object_type:
        filtered_objects = [obj for obj in filtered_objects if obj.type == object_type]
    
    if month:
        filtered_objects = [obj for obj in filtered_objects if obj.month == month]
    
    if hemisphere:
        filtered_objects = [obj for obj in filtered_objects 
                           if obj.hemisphere == hemisphere or obj.hemisphere == 'Both']
    
    return [obj.to_dict() for obj in filtered_objects]